gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -279,11 +285,26 @@ private GoogleCloudStorageItemInfo getObjectMetadata(StorageResourceId resourceI
         : new IOException("Error reading " + resourceId, exception);
   }
 
+  private void recordGCSAPILatency(
+      MeasureLong measure, long time, String status, String method, String protocol) {
+    if (storageOptions.isTraceLogEnabled()) {
+      Map<String, Object> jsonMap = new HashMap<>();
+      jsonMap.put("method", method);
+      jsonMap.put("status", status);
+      jsonMap.put("protocol", protocol);
+      jsonMap.put("duration", time);
+      jsonMap.put("measure", measure);
+
+      logger.atInfo().log("%s", LazyArgs.lazy(() -> gson.toJson(jsonMap)));
+    }
+  }
+
   private void recordSuccessMetric(
       MeasureLong measure, Stopwatch stopwatch, String method, String protocol) {
     long time = stopwatch.elapsed(MILLISECONDS);
     TagKey[] keys = new TagKey[] {METHOD, STATUS, PROTOCOL};
     String[] values = new String[] {method, STATUS_OK, protocol};
+    recordGCSAPILatency(measure, time, STATUS_OK, method, protocol);
     metricsRecorder.recordLong(keys, values, measure, time);
     logger.atFinest().log(
```

Comment:
why not convert this into a JSON and log, rather than adding another log statement? Now the same information will be logged twice.

Reply:
Because we also want to control the trace logging with flag.